tensorflow_quantum/core/ops/math_ops/inner_product_op.py

```diff
@@ -82,8 +129,19 @@ def inner_product(programs, symbol_names, symbol_values, other_programs):
         `tf.Tensor` with shape [batch_size, n_others] where `out[i][j]` is equal
             to the inner product of `programs[i]` with `symbol_values[i]`
             resolved in and `other_programs[i][j]`.
-
     """
+
+    def grad(dy):
+
+        def _true_grad():
+            return _inner_product_adj_grad(programs, symbol_names,
+                                           symbol_values, other_programs, dy)
+
+        inner_prod_grad = tf.cond(tf.math.equal(symbol_names.shape[0], 0),
+                                  lambda: tf.zeros_like(symbol_values),
+                                  _true_grad)
```

Comment:
I don't think `symbol_names.shape[0]` is accesible in non-eager mode. Would you be able to change this so that it looks more like: https://github.com/tensorflow/quantum/blob/master/tensorflow_quantum/python/differentiators/differentiator.py#L36 ?
Where we use `tf.size` and `tf.equal` instead ?

Reply:
Done.